### ISSUE TITLE
[with-react-intl]: polyfills don't seem to work on first page load

### ISSUE THREAD

**Issue by melanieseltzer** [vercel/next.js#20311]
# Bug report

## Describe the bug

I pulled down the [with-react-intl example](https://github.com/vercel/next.js/tree/canary/examples/with-react-intl) to test it out in Safari with the [ListFormat](https://formatjs.io/docs/polyfills/intl-listformat) polyfill added into the polyfills.ts file (following the same format as the rest). 

I had been trying to integrate the polyfills (with ListFormat) into a separate proprietary app but it was not working for me... so that's when I decided to test out the example as well, to make sure I wasn't doing anything wrong.

It appears that something is a bit wrong with the example, since it always errors on page load, but the polyfills kick in when doing client-side navigation and work fine after that.

How do I make it work so it polyfills on page load? I'm assuming it has something to do with it getting called in getInitialProps which would be called on the server first and then on the client when using Link, right?

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Use the [with-react-intl example](https://github.com/vercel/next.js/tree/canary/examples/with-react-intl) example to create a new next project
2. Add [ListFormat](https://formatjs.io/docs/polyfills/intl-listformat) polyfill into the polyfills.ts file
3. Run the dev server (dev-no-custom-server) and open up in Safari
4. See error on page load

## Expected behavior

It polyfills on the client on first load.

## Screenshots

![Kapture 2020-12-18 at 11 14 43](https://user-images.githubusercontent.com/17421347/102652035-51ae4100-4122-11eb-9fae-72f384a2634d.gif)

## System information

- OS: macOS
- Browser (if applies) Safari (to see the bug, since it doesn't support Intl.ListFormat)
- Version of Next.js: latest
- Version of Node.js: v10.16.0
- Deployment: n/a

**Comment by melanieseltzer** [comment]
cc @longlho

**Comment by longlho** [comment]
how did u add the polyfill?

**Comment by melanieseltzer** [comment]
@longlho Here's a commit where I added it (just a small repo to show the bug, pretty much kept everything as-is from the example, except fixed the lang path): https://github.com/melanieseltzer/next-react-intl-app-polyfill-bug/commit/7a3bca2ee2b4291a441e3bbda5d05c82a91161a8

**Comment by longlho** [comment]
Have you tried using latest versions of everything?

**Comment by melanieseltzer** [comment]
Yes, just tried bumping all the packages, although most are not severely out of date (because I generated the example using `npx create-next-app --example with-react-intl`).

The only thing out of date is Node, but the reason is because I'm ultimately trying to integrate the with-react-intl example into a proprietary app (that I can't share publicly) and its Node version is locked to 10.16.0 (not sure if we can upgrade but assume we can't for now).

However, I just did a test using 14.15.3, and I get the same problem in the browser. It's polyfilling just fine on the server:

14.15.3 - detects Intl.Locale needs polyfill
10.16.0 - detects both Intl.Locale and Intl.ListFormat needs polyfill, adds the listformat locale data

This works fine when the app is booted up, but yeah... as soon as I actually go into Safari, it suddenly is not working, until you do some client-side navigation.

**Comment by melanieseltzer** [comment]
From the Next docs:

> For the initial page load, getInitialProps will run on the server only. getInitialProps will then run on the client when navigating to a different route via the next/link component or by using next/router.

So, for initial load, there is indeed no polyfilling on the client at all 🤔  Does it need to be added in say, a useEffect also?

**Comment by longlho** [comment]
Hmm looks like with the dev-no-custom-server that's the behavior. Tbh I'm not a nextjs expert so your guess is as good as mine :) you can prob give useEffect a try.

**Comment by exogen** [comment]
Unfortunately `useEffect` won't fix it either, as the data needs to be there on first render (hydration) in order to match the server output.

As the example is written, this is (to my knowledge) unsolvable in Next.js at the moment, since it doesn't offer an affordance for awaiting something asynchronously before hydration occurs.

The way react-intl integration worked previously was better for Next.js because you added the locale data for the given locale to the `_document`, which allowed you to then synchronously `addLocaleData` using it before the app hydrated. In theory you could still add the literal `locale-data/{lang}.js` script content into a `<script>` tag in `_document` so it'd be just like old times.

Your other option is to use regular non-dynamic `import` in `_app` on all the possible locales, but it'd be loading the locales you don't need unnecessarily.

**Comment by floriangosse** [comment]
Does someone have found a solution or workaround for that problem? The only solution I've in mind is to load the polyfills synchronously for every client and locale.

**Comment by melanieseltzer** [comment]
@floriangosse I ended up rotating off the project that prompted this initial report, so I'm afraid I didn't look into it further 😅 Do you want to post an example with the solution you mentioned? Maybe it would be helpful for folks here.

**Comment by floriangosse** [comment]
I came up with an even simpler solution – but still a hacky one: I use Node.js 14 which has full internationalization support (it's in Node.js since version 13) so that means I only have to load the polyfills on the client side.

```
if (typeof window !== 'undefined') {
    // NOTE: This is necessary because Next.js doesn't provide a way to run a function asynchronously – so we have to load
    // the Intl polyfills synchronous.
    appLocales.forEach((appLocale) => {
        polyfillSync(appLocale);
    });
}
```

As you can see, right now I load the polyfills for EVERY locale as a language change can happen during the runtime and this was the easiest solution right now. I even implemented a synchronous version of my polyfill function that uses synchronous `require` calls to load the necessary polyfills.
